File: content/problems/3116-kth-smallest-amount-with-single-denomination-combination/solutions/binary-search-with-inclusion-exclusion/Solution.java
class Solution {
    private long gcd(long first,long second){
        while(second!=0){
            long value=first%second;
            first=second;
            second=value;
        }
        return first;
    }
    private long count(int[] coins,long limit){
        long total=0;
        for(int mask=1;mask<(1<<coins.length);mask++){
            long lcm=1;
            int bits=0;
            boolean tooLarge=false;
            for(int index=0;index<coins.length;index++)if((mask&(1<<index))!=0){
                bits++;
                long divisor=gcd(lcm,coins[index]);
                if(lcm/divisor>limit/coins[index]){
                    tooLarge=true;
                    break;
                }
                lcm=lcm/divisor*coins[index];
            }
            if(!tooLarge){
                long occurrences=limit/lcm;
                total+=(bits&1)==1?occurrences:-occurrences;
            }
        }
        return total;
    }
    public long findKthSmallest(int[] coins,int k){
        int minimum=Arrays.stream(coins).min().getAsInt();
        long low=1,high=(long)minimum*k;
        while(low<high){
            long middle=low+(high-low)/2;
            if(count(coins,middle)>=k)high=middle;
            else low=middle+1;
        }
        return low;
    }
}

File: content/problems/3116-kth-smallest-amount-with-single-denomination-combination/solutions/binary-search-with-inclusion-exclusion/solution.cpp
class Solution {
    long long gcdValue(long long first,long long second){
        while(second){
            long long value=first%second;
            first=second;
            second=value;
        }
        return first;
    }
    long long countAmounts(vector<int>& coins,long long limit){
        long long total=0;
        for(int mask=1;mask<(1<<(int)coins.size());++mask){
            long long lcm=1;
            int bits=0;
            bool tooLarge=false;
            for(int index=0;index<(int)coins.size();++index)if(mask>>index&1){
                ++bits;
                long long divisor=gcdValue(lcm,coins[index]);
                if(lcm/divisor>limit/coins[index]){
                    tooLarge=true;
                    break;
                }
                lcm=lcm/divisor*coins[index];
            }
            if(!tooLarge){
                long long occurrences=limit/lcm;
                total+=bits&1?occurrences:-occurrences;
            }
        }
        return total;
    }
    public:long long findKthSmallest(vector<int>& coins,int k){
        long long low=1,high=1LL*(*min_element(coins.begin(),coins.end()))*k;
        while(low<high){
            long long middle=low+(high-low)/2;
            if(countAmounts(coins,middle)>=k)high=middle;
            else low=middle+1;
        }
        return low;
    }
};
